src/Cardano/Wallet/CoinSelection/LargestFirst.hs

```diff
@@ -0,0 +1,111 @@
+{-# LANGUAGE RankNTypes #-}
+
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- This module contains the implementation of largestFirst
+-- input selection algorithm
+
+
+module Cardano.Wallet.CoinSelection.LargestFirst (
+    largestFirst
+  ) where
+
+import Prelude
+
+import Cardano.Wallet.CoinSelection
+    ( CoinSelection (..), CoinSelectionError (..), CoinSelectionOptions (..) )
+import Cardano.Wallet.Primitive.Types
+    ( Coin (..), TxIn, TxOut (..), UTxO (..), balance )
+import Control.Monad
+    ( foldM, when )
+import Control.Monad.IO.Class
+    ( MonadIO )
+import Control.Monad.Trans.Except
+    ( ExceptT (..), throwE )
+import Data.List.NonEmpty
+    ( NonEmpty (..) )
+import Data.Ord
+    ( comparing )
+
+import qualified Data.List as L
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map.Strict as Map
+
+
+-- | Largest-first input selection policy
+largestFirst
+    :: forall m. MonadIO m
+    => CoinSelectionOptions
+    -> UTxO
+    -> NonEmpty TxOut
+    -> ExceptT CoinSelectionError m CoinSelection
+largestFirst opt utxo txOutputs = do
+    let txOutputsSorted = NE.toList $ NE.sortBy (flip $ comparing coin) txOutputs
+    let n = fromIntegral $ maximumNumberOfInputs opt
+    let nLargest = take n . L.sortBy (flip $ comparing (coin . snd)) . Map.toList . getUTxO
+    let moneyRequested = sum $ (getCoin . coin) <$> txOutputsSorted
+    let utxoBalance = fromIntegral $ balance utxo
+    let numberOfUtxoEntries = fromIntegral $ L.length $ (Map.toList . getUTxO) utxo
+    let numberOfTransactionOutputs = fromIntegral $ NE.length txOutputs
+
+    when (utxoBalance < moneyRequested)
+        $ throwE $ NotEnoughMoney utxoBalance moneyRequested
+
+    when (numberOfUtxoEntries < numberOfTransactionOutputs)
+        $ throwE $ UtxoNotEnoughFragmented numberOfUtxoEntries numberOfTransactionOutputs
+
+    case foldM atLeast (nLargest utxo, mempty) txOutputsSorted of
+        Just (_, s) ->
+            return s
+        Nothing ->
+            throwE $ MaximumInputsReached (fromIntegral n)
+
+
+-- Selecting coins to cover at least the specified value
```

Comment:
Added this extensive comment how the largestFirst works - hope it will save time anyone trying to understand it